### ISSUE TITLE
Apps next steps (Rome edition)

### ISSUE THREAD

**Issue by mariusandra** [PostHog/meta#63]
Another hackathon has passed and apps have gotten significantly more powerful:

![2022-09-27 19 41 48](https://user-images.githubusercontent.com/53387/192598538-65c70c48-fc7d-40af-9d72-d2b468e77339.gif)

Here's the todo list, first order:

- [x] One more chat about naming things. "Web inject apps"? "web.ts"? **Winner: [Site apps](https://github.com/PostHog/posthog/pull/12179)**
- [ ] Figure out a CDN/caching strategy. What we have now is fine for internal testing and usage on posthog.com and the app, but it seems like the [Rome hackathon blog post](https://posthog.com/blog/rome-hackathon) might have raised the expectations on what people expect here. **Progress documented below**
- [ ] Create a great app authoring experience (part 1, documentation) **documented below in this issue, needs to be updated for [site.ts](https://github.com/PostHog/posthog/pull/12179) and added to the docs**
  - [ ] Document `opt_in_site_apps` for posthog-js.
  - [ ] Document `site.ts` and its `export function inject({ config, posthog }) {}`
  - [ ] Document `site: true` in `config` in `plugin.json`
  - [ ] Document [web-dev-server](https://github.com/PostHog/web-dev-server) and how it can be used to [run site.ts apps](https://github.com/PostHog/feedback-app/blob/main/package.json#L8) locally.
  - [ ] Create a site.ts app template repo and document a flow with `npx degit` to get started ([example](https://keajs.org/docs/intro/installation#shortcut-1-vite-starter)).
- [ ] Build and publish useful apps
  - [x] [Pineapple mode](https://github.com/PostHog/pineapple-mode-app)
  - [x] [Notification bar](https://github.com/PostHog/notification-bar-app)
  - [x] [Feedback widget](https://github.com/PostHog/feedback-app)
  - [ ] NPS surveys

Here are the less urgent todos:

- [ ] Create a great app authoring experience (part 2, new stuff - TODO **later**)
  - [ ] Split out app transpilation into a separate service/flow, instead of the current hacky plugin server setup and transpilation step.
  - [ ] Display in the interface if a plugin is compiled/transpiled or not
  - [ ] Clearly distinguish between "site", "frontend-sidebar", "backend-process-event", "backend-export", etc apps in the interface and show relevant warnings (e.g. in the apps interface, we should suggest to opt in to injection if not done so, when installing a site app)
  - [ ] Technically you could edit `site.ts` on cloud, since this is not code we run in the backend. Investigate if this is something we could feasibly enable. 
  - [ ] Create an app source edit history. This may be needed for the CDN anyway to avoid URLs that break?
- [ ] Finish frontend apps **(we can still punt on this)**
  - [x] [Create a PR](https://github.com/PostHog/posthog/pull/12257) to remove the flag. Small changes in the sidebar and apps editor ensue. We can launch this.
  - [ ] Document `frontend.ts`
  - [ ] Document `@posthog/lemon-ui` and `@posthog/apps-common`. 
  - [ ] Create a similar local dev server to site.ts apps
- [ ] Security
  - [ ] Add 2FA support.
  - [ ] Lock down who can edit apps. E.g. only certain users, only 2FA enabled, etc.
  - [ ] Talk through the implications enabling this on app.posthog.com (done) brings to self hosted users (I'd enable these apps only on cloud, and disable for self-hosted?).
- [ ] [Rename "Launch Toolbar" to "My Sites".](https://github.com/PostHog/posthog/issues/12181) For each site show:
  - [ ] [Screenshot](https://github.com/PostHog/posthog/issues/12082)
  - [ ] What posthog-js version it's using
  - [ ] Are site apps enabled or not
  - [ ] Other handy statistics
  - [ ] Button to launch toolbar

I'll create a few PRs from this.

**Comment by mariusandra** [comment]
Brain-dump of the state of things.

### Naming things

I've been using the terms "frontend app" and "web app" and "injection app" and "js app" rather interchangeably. We need to be clear on how we call things to communicate effectively between ourselves and to the "app developers".

We now have three separate types of apps:

- Ingestion apps (plugins? name pending)
  - These are the original plugins.
  - They hook into the ingestion lifecycle somewhere, or run parallel to it. 
  - Depending on the functions your app's `index.ts` file exports, these apps expose different `capabilities` like `processEvent`, `exportEvents`, `runEveryMinute`, tasks, jobs, etc. 
  - All of these apps are run on the backend on our infrastructure. Thus we must be very careful to only run trusted code here, otherwise someone could set up a bot farm or whatnot. This is why we haven't enabled app editing on cloud yet - too risky.

- Sidebar apps (frontend apps? name pending)
  - These are written by exporting a React component in `frontend.tsx`.
  - These are transpiled from TS to JS in our backend, and then run on the client side in the browser. Thus we could theoretically allow editing of these kinds of apps on cloud - no risk of malicious code on our backend. Risks do remain on the frontend.
  - They're limited to just calling the existing PostHog API, rendering a React component and using Kea logics (no typegen though). There's no place for "app data" other than finding and editing the plugin's config via the API, or [hijacking feature flag variant json](https://github.com/PostHog/notification-bar-app/blob/f258ee29f67ae97aa76b972d2c4210e173fa5548/frontend.tsx).
  - They can import components and utils from [`@posthog/apps-common`](https://github.com/PostHog/posthog/blob/master/frontend/%40posthog/apps-common/src/index.ts) and [`@posthog/lemon-ui`](https://github.com/PostHog/posthog/blob/master/frontend/%40posthog/lemon-ui/src/index.ts). These are packages that actually exist and can be used for local development, though we haven't published new versions in a while.

- Site apps. (on-site apps? web inject apps? in-site apps? name pending)
  - These are the new ones. Pineapple mode, feedback widget, etc.
  - You write code in `web.ts` and if you have initialized posthog-js with `posthog.init('key', {opt_in_web_app_injection: true})`, this same code will run on your site.
  - Example apps: [pineapple mode](https://github.com/PostHog/pineapple-mode-app/blob/main/web.ts), [feedback widget](https://github.com/PostHog/feedback-app/blob/main/web.ts).
  - These apps **do not** have access to React: you must write your UI from scratch using `document.createElement('div')`. We might offer this one day, but it's outside of the scope for now: let's get the core solid first.

#### Actions
- I propose calling these three categories:
  - "sidebar apps" - and renaming `frontend.tsx` to `sidebar.tsx`
  - "site apps" - and renaming `web.ts` to `site.ts`
  - "ingestion apps" - and keeping it in `index.ts` for now

Anyone against?

**Comment by mariusandra** [comment]
### Security concerns
   
The TLDR security concern with apps is this:
- If you have manually enabled `opt_in_web_app_injection` when loading the PostHog JS library, then anyone who has access to your PostHog project/org can theoretically add arbitrary code to your app's frontend.

This also means any admin of PostHog with access to impersonate users (e.g. when support hero), can theoretically inject code onto your website.

The messaging around this should be simple: site apps are a beta feature with certain risks (what's above). It's not yet ready for a general rollout. You should opt in only if you understand and agree with the risks.

To mitigate these risks, we need to 1) build a 2FA system (mitigated if you're using SSO/SAML), 2) improve our own admin/impersonate security (needed anyway for SOC2), 3) build granular permissions around who can build and who can install/configure apps.

### Usage steps

1. `opt_in_web_app_injection` for posthog-js.

To enable site apps, you must manually opt in:

> ```ts
> posthog.init("api_key", { 
>     "api_host": "https://app.posthog.com",
>     "opt_in_web_app_injection": true 
> })
> ```

Q: Should we rename this to `enable_site_apps: true`?

2. You must create a file `web.ts` and put all app code there.

Then write the following code:

```ts
export function inject({ config, posthog }) {

}
```

- `config` gets the value of every app config parameter from `plugin.json`, which has `web:true` set.
- `posthog` is a copy of the posthog instance (e.g. run `posthog.capture`)

3. Inside `plugin.json`, specify `web:true` for data you want to share to your site:

```json
{
    "name": "Feedback Widget",
    "config": [
        {
            "markdown": "## MANUAL STEP NOTICE: This app needs to injects code into your website through posthog-js. You need to **opt-in** on your site to enable this behaviour.\n\n```\nposthog.init(\"api_key\", {\n  \"api_host\": \"https://app.posthog.com\",\n  \"opt_in_web_app_injection\": true,\n})\n```"
        },
        {
            "key": "buttonText",
            "name": "Button text",
            "type": "string",
            "default": "🥹",
            "required": true,
            "web": true
        },
        {
            "key": "apiKey",
            "name": "API KEY",
            "type": "string",
            "required": true,
        }
    ]
}
```

Here the `apiKey` config parameter will not be exposed to the site app.

4. There's [`@posthog/web-dev-server`](https://github.com/PostHog/web-dev-server), which can be used to [run web.ts apps](https://github.com/PostHog/feedback-app/blob/main/package.json#L8) locally.

This will expose a local server on port `3040`:

<img width="601" alt="image" src="https://user-images.githubusercontent.com/53387/194887907-c9d6163f-328f-48b6-a083-933b4e4e6ccb.png">

... which can be used to develop the app locally for faster cycle times:

<img width="508" alt="image" src="https://user-images.githubusercontent.com/53387/194888017-180f8adc-f8c2-4337-8955-5623b094b9e2.png">


### Template repo

We should also create a template repo or SDK to make local app development easier. Currently the best instructions I can provide are these:

```
npx degit https://github.com/PostHog/feedback-app my-app
cd my-app
npm install
npm start
```

This is not *that* bad for getting started to be honest.

**Comment by tiina303** [comment]
I like the idea of separating apps more clearly. A bit out of scope for this issue here maybe, but I would break up ingestion apps up more too (could be in the same file 🤷‍♀️ ). With CDP direction in mind it could be:
1. transform apps - i.e. stuff that has `processEvent` - one likely shouldn't combine this with other functions, we have different requirements for these apps, specifically this needs to be fast as it's within the ingestion path before we have written the event to CH.
2. destination apps: `exportEvents`, `onEvent`, `runEveryMinute`, ...
3. source apps: `runEveryMinute`, ...

**Comment by mariusandra** [comment]
Some details on one more topic:

### CDN/caching strategy. 

All injected [site apps](https://github.com/PostHog/posthog/pull/12179) are loaded from an URL like:

```
https://app.posthog.com/web_js/8/IVFwPfyqPAh9v9plTkYqzJ69z14lgXEr3eRo411vXTd/
```

The format is `/web_js/:plugin_config__id/:plugin_config__web_token/`

These urls/paths are received by posthog-js from the call to `/decide`. Each hit to this `/web_js` url makes one query to load the transpiled JS from postgres, and another one to read data from the app's config rows with `web: true`.

This is begging for a CDN and unique URLs with a very long TTL.

The simplest solution for now is to append a unique `/:hash/` at the end of the URL, calculated from e.g. `md5(plugin_config.updated_at + '🥹' + plugin_source_file.updated_at)`. The API would just ignore this hash and always serve the latest content. 

A more advanced version keeps a log of all plugin source and config changes, and for the same hash, always serves the same source/config to avoid flakiness that might happen between a delay in loading `/decide` and the site apps, which may have changed in between. If we always return the latest content at the endpoint, and pass it through a CDN, this will probably be a non issue.

**Comment by weyert** [comment]
Would it be possible to disable this on a project level on Posthog Cloud?
